### ISSUE TITLE
Extend /ignore system with pattern matching

### ISSUE THREAD

**Issue by c22** [thelounge/thelounge#2573]
Continuing on from #464 and #2197  it would be great if we could have more flexibility when adding ignores.

Particularly if we could implement regular expression/pattern matching of specific messages to ignore, such as ignoring messages containing specific words or ignoring CTCP messages.

**Comment by RockyTV** [comment]
As the author of the feature, I'm not sure if `/ignore` would be a good idea to ignore messages. The purpose of the command is to "block" messages from a certain user/mask from showing up. If there's a feature that notifies you when certain words are mentioned, this could be reworked to implement support for a ignored keywords, and then you could have pattern matching to ignore messages that contain the specified keywords.

TL;DR: IMO `/ignore` should be a command to block users, not messages

**Comment by dgw** [comment]
Pattern matching the message content would be complicated, to be sure, but… What about message types? Like, ignoring highlights from a user but still being able to see their messages, or blocking PMs from them but still seeing what they say in channels, etc. Basically, some or all of the subcategories shown in my OP from #464 (reproduced below):

![Textual 5 /ignore window](https://cloud.githubusercontent.com/assets/164140/16600850/3a7d141e-42d7-11e6-8dbf-8c0689ba77c3.png)

**Comment by c22** [comment]
I admit to not being an IRC expert. I had assumed these things were all raw messages in some form or another which could be pattern matched (from my reading, CTCP is just a PRIVMSG surrounded by ASCII value 0x01 characters).

I envisaged a system where you could ignore a message like:
`:nick!username@somehost.com PRIVMSG c22 :CTCP c22 CTCP VERSION`
with a generic pattern like:
`:[^!]+![^@]+@.* PRIVMSG c22 :\x01.*`

Aside from highlights, aren't all the options specified by @dgw implemented by raw messages in some way or another?

**Comment by dgw** [comment]
> Aside from highlights, aren't all the options specified by @dgw implemented by raw messages in some way or another?

They are all properties of raw IRC messages, yes, but by the time The Lounge is handling messages they're not raw lines any more. The IRC framework does the parsing and passes along message objects, not plain text. (Also, as @McInkay said, having filters on raw lines would be a UX nightmare.)

**Comment by c22** [comment]
It's a good UX for power users :)

**Comment by McInkay** [comment]
> It's a good UX for power users :)

That's not how UX works.

It's not good UX full stop. You may think you want that feature, but let's be realistic, trying to remember all the different regex you will have to put to block a certain type of message becomes less than fun.

>  we should change how messages are ignored. At the moment, ignored messages are just dropped, which messages sent prior to ignoring will still show up, and messages sent after un-ignoring will never show up.

Agreed, I think this is where we should go. Mark them as "ignored" and hide them in the client.

I'll open a new issue for that specifically.

**Comment by c22** [comment]
> It's like saying a feature is accessible for sighted people :)

You mean like syntax highlighting?

> That's not how UX works.

Not sure what you mean, UX is user experience, it's completely subjective based on who your users are. Good UX doesn't mean a good experience for everyone, it means a good experience for your users. Currently your users are people who want to run a client/server IRC setup (people like me are your users). They're not special snowflakes who have no idea what a regex is. I agree it's less than fun but setting up thelounge how I like it has been less than fun (on the plus side I only need to do it once) . If I wanted a "fun" chat client I'd be using Slack.

Moving on to topics more fruitful, I really would like the ability to block CTCP messages as they are a privacy risk. If that should not be implemented as part of /ignore then what should be done? There are already things that can only be done by config files, I don't particularly care if this is one of those things.

Give some direction and maybe I can work on a PR to get it done.

**Comment by xnaas** [comment]
Just commenting to add that I'd like something like the regex support of [the Ignore module from ZNC](https://wiki.znc.in/Ignore). Someone was saying that makes more sense for it to be a plugin when those are a thing, rather than core functionality. I can agree with that to an extent.
